server/graphql/v1/types.js

```diff
@@ -313,7 +316,10 @@ export const MemberType = new GraphQLObjectType({
       member: {
         type: CollectiveInterfaceType,
         resolve(member, args, req) {
-          return member.memberCollective || req.loaders.collective.findById.load(member.MemberCollectiveId);
+          const memberCollective =
+            member.memberCollective || req.loaders.collective.findById.load(member.MemberCollectiveId);
+          memberCollective.inTheContextOfCollectiveId = member.CollectiveId;
```

Comment:
This doesn't "smell" good.

Reply:
Not exactly sure what to do with this feedback ¯\\\_(ツ)_/¯ 

The goal here is to keep track of the context of the collective in which this user is.
As an admin of the collective, I should only be able to access the email address of an anonymous backer in the context of that collective. Otherwise the permission system would give me access if the same anonymous user made another donation to another collective that I am an admin of.

Reply:
The pattern is not good. Modifying an object on the fly like that will be source of bugs.

For example, what if what you get out of the resolver is not a copy but a shared object and you end up sending the modified version in unexpected places?

The removal of `createdByUser` in transactions could be a simpler solution.

Reply:
It's used in the REST api. 
https://github.com/OpenCollective/opencollective-api/blob/3c1cbfd4a72b3d19987cae33559a135fc5fa3cef/server/routes.js#L219
https://github.com/OpenCollective/opencollective-api/blob/3c1cbfd4a72b3d19987cae33559a135fc5fa3cef/server/graphql/v1/restapi.js#L23-L26

I don't know if anyone is using it. So what do we do?

> For example, what if what you get out of the resolver is not a copy but a shared object and you end up sending the modified version in unexpected places?

I don't understand. This is not going to interfere with anything else. And if it's missing it falls back to the safe scenario which is not sending the personal data.

Reply:
If it's only used in the REST API, it's already a very limited usage (I only have 2 users in mind), and very likely no one is relying on the specific field. Not a blocker for me.

Reply:
Isn't webpack relying on getting the email addresses of their backers for their voting platform https://webpack.js.org/vote/ ?

Reply:
No trace in the logs, I only see Triplebyte using it.

Reply:
Well, one too much. I could see why they would want to get the email addresses of the people using the gift card.

Then maybe we could simplify by not returning the `Transaction.createdByUser` when it's anonymous even if you are an admin?

But it's weird that only in that case we don't return it. It feels like a bug.

Reply:
Can you ask them if they rely on that feature?

Reply:
I'm sure they don't rely on it because this functionality was broken until February and they never complained. https://github.com/opencollective/opencollective-api/commit/8b36cc52ba15b5b081b6263c355ebe90048401dc

Reply:
As such the `inTheContextOfCollectiveId` really looks like a hack. It's not documented, not enforced in the model and you have to go through the code to understand what it does. That will confuse any external contributors, or us if we look at it in 2 months.

@znarf's solution looks good to me.

But if we really want to go with the current pattern, we should at least make it explicit so we don't create new technical debt. We could do this by adding the field to the model (`DataTypes.VIRTUAL`) with a proper description.

Reply:
Would it make more sense to have that variable in the context of the resolver (the 3rd argument `req`)?

We have to return the `createdByUser` when the host is logged in.

Reply:
> Would it make more sense to have that variable in the context of the resolver (the 3rd argument req)?

Yes

Reply:
> We have to return the createdByUser when the host is logged in.

I'm not following. What is the relation with the current discussion?

Also, my feedback to move forward simple and efficient, deprecate `createdByUser` in transactions and send a fake object in it **all the time**.

Reply:
This is in the context of getting the `createdByUser` based on the `memberCollectiveId` which references the incognito profile.

But from GraphQL's perspective, it doesn't matter much what the path is.

I tried but we can't keep track of the context in which a memberCollective is in the context of the resolver because this could be overwritten if we browse members from multiple collectives.

So look, I know it's a bit hacky but it works and it has tests.
I can't come up with anything better for now with the constrain that we have to return the personal info of the incognito profile if logged in as an admin of the host.

I'll add a `DataTypes.VIRTUAL` as suggested by @Betree .

I'm trying my best to help you guys so that you don't have to build this particular feature but I'm not full time on the code and I'm not as talented as you are. So I can help, but getting to your level of code quality would require a different level of commitment and focus on the code that I just can't offer right now. Thank you for your understanding.